### PR TITLE
Document renumbering lines in macros

### PR DESCRIPTION
**Pull request by bramtayl**
So, this is something I've figured out on my own, so I'm not 100% sure if it's correct or not. If it's not correct, I'd greatly appreciate knowing what the correct way to handle this is.

### CODE REVIEW

File: doc/src/manual/metaprogramming.md
Comment (imciner2) [inline doc/src/manual/metaprogramming.md:698]: I think it would be better to link to the "macro invocation" section where the `__source__` arg is defined rather than just saying "above", since that isn't too descriptive.

File: doc/src/manual/metaprogramming.md
Comment (imciner2) [inline doc/src/manual/metaprogramming.md:677]:
```suggestion
When a macro is invoked, Julia will introduce a new `LineNumberNode` that points to to the definition expression inside the macro.
```

File: doc/src/manual/metaprogramming.md
Comment (imciner2) [inline doc/src/manual/metaprogramming.md:681]:
```suggestion
- If an error occurs in code introduced in a macro, you likely want stack-traces to point to the user's call site, rather than the macro's internals.
```

File: doc/src/manual/metaprogramming.md
Comment (imciner2) [inline doc/src/manual/metaprogramming.md:682]:
```suggestion
- If a user calls a macro, you likely want coverage to consider the user's call site as covered, rather than the macro itself.
```

File: doc/src/manual/metaprogramming.md
Comment (imciner2) [inline doc/src/manual/metaprogramming.md:698]:
```suggestion
`__source__` will contain a `LineNumberNode` pointing to the user's call site (see more information about `__source__` above).
```

File: doc/src/manual/metaprogramming.md
Comment (imciner2) [inline doc/src/manual/metaprogramming.md:699]:
```suggestion
The macro will now renumber the new line of code to point the user's call site, rather than inside the macro.
```

**Review by imciner2 — CHANGES_REQUESTED** [review]
I found the term "definition expression" somewhat confusing in this because it actually isn't defined elsewhere in the docs and so was a new term. Instead I think it is simpler to talk about the macro itself in these cases.

**Comment by bramtayl** [comment]
Thanks for the suggestions! I was having trouble coming up with good terminology.

File: doc/src/manual/metaprogramming.md
Comment (imciner2) [inline doc/src/manual/metaprogramming.md:699]:
```suggestion
`__source__` will contain a `LineNumberNode` pointing to the user's call site
(see [Macro invocation](@ref) for more information about `__source__`).
```

Minor change, since the opening parenthesis was missing.

**Comment by imciner2** [comment]
Yea, choosing terminology is always a difficult thing to do. I noticed one last thing that needs to be cleaned up (the parenthesis was missing), and after that it looks good to me (but you need to wait for someone with commit rights to approve this as well).

**Review by imciner2 — APPROVED** [review]

File: doc/src/manual/metaprogramming.md
Comment (vtjnash) [inline doc/src/manual/metaprogramming.md:682]: These argument seem inaccurate to me. The __source__ should already be be getting represented as the place where the macro is being inlined at (as a fake "caller"). Injecting it again into the macro body seems like it could lead to even stranger frames as it tries to guess why this macro is changing back and forth between inlined and local. If the source location is missing in the caller (where we splice the macro to), that seems like just a bug.

There's already `Base.remove_linenums!` and associated improvements in PR form: https://github.com/JuliaLang/julia/pull/31335 and `Meta.replace_sourceloc!` from https://github.com/JuliaLang/julia/pull/22623.

Comment (bramtayl) [reply doc/src/manual/metaprogramming.md:682]: I honestly understood very little of that comment...

Comment (bramtayl) [reply doc/src/manual/metaprogramming.md:682]: > If the source location is missing in the caller (where we splice the macro to), that seems like just a bug.

Ok, if I understand you right, a bug like this is likely what I'm running into

Comment (bramtayl) [reply doc/src/manual/metaprogramming.md:682]: Should I close this and open a new issue?

**Comment by bramtayl** [comment]
Replaced by https://github.com/JuliaLang/julia/issues/41043

**Pull request closed without merging**
